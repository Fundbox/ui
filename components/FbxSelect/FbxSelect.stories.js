import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withInfo } from 'storybook-addon-vue-info';
import { array } from '@storybook/addon-knobs';

import FbxSelect from './FbxSelect.vue';
import FbxButton from './../FbxButton/FbxButton.vue';
import summary from './FbxSelect.md';
const stories = storiesOf('Select', module);
const withSummary = withInfo({ summary });

const data = () => ({
  selected: "",
  options: array('Options', ['Option 1', 'Option 2', 'Option 3', 'Option 4']),
});

const watch = {
  selected (value) {
    action(`New selected value: ${value}`)()
  },
};

const defaultStory = {
  components: { FbxSelect },
  data,
  watch,
  template: `
  <div>
    <fbx-select v-model="selected">
      <option value="">Default</option>
      <option v-for="option in options" :value="option">{{ option }}</option>
    </fbx-select>
    <br/>
    <h3>Selected option: {{selected}}</h3>
  </div>
  `,
};
stories.add('default', withSummary(() => defaultStory));

const validationsStory = {
  components: { FbxSelect, FbxButton },
  data,
  watch,
  methods: {
    handleSubmit () {
      this.$validator.validate().then(valid => {
        if (valid) {
          action(`Submitted value: ${this.selected}`)()
        } else {
          action("Form is invalid. Submittion prevented.")()
        }
      });
    }
  },
  template: `
  <form @submit.prevent="handleSubmit">
    <fbx-select name="select" v-model="selected" validations="required">
      <option value="">Default</option>
      <option v-for="option in options" :value="option">{{ option }}</option>
    </fbx-select>
    <br/>
    <fbx-button type="submit">Submit</fbx-button>
    <br/>
    <h3>Selected option: {{selected}}</h3>
  </form>
  `,
};

stories.add('validations', withSummary(() => validationsStory));
