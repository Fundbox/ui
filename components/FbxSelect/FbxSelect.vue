<template>
  <div>
    <select
      class="fbx-select"
      v-validate="validations"
      v-bind="$attrs"
      v-on="listeners"
      :value="value"
      :class="{ invalid: isInvalid }"
    >
      <slot></slot>
    </select>
    <fbx-validation-message class="validation-message" v-if="isInvalid">{{ validationMessage }}</fbx-validation-message>
  </div>
</template>

<script>
import FbxValidationMessage from "../FbxValidationMessage/FbxValidationMessage.vue";
export default {
  name: "FbxSelect",
  inheritAttrs: false,
  inject: ["$validator"],
  components: {
    FbxValidationMessage
  },
  props: {
    validations: String,
    value: String
  },
  computed: {
    isInvalid() {
      return this.errors.has(this.$attrs.name);
    },
    validationMessage() {
      return this.errors.first(this.$attrs.name);
    },
    listeners() {
      return {
        // Pass all component listeners directly to button
        ...this.$listeners,
        change: ({ target }) => {
          this.$emit("input", target.value);
        }
      };
    }
  }
};
</script>

<style lang="scss" scoped>
@import "./../styles/utils/color-palette";

.fbx-select {
  font-size: inherit;
  font-weight: inherit;
  font-family: inherit;
  -webkit-appearance: none;
  appearance: none;
  padding: 15px;
  color: $extra-dark-gray;
  background: url("../assets/arrow-down.svg") no-repeat right 10px center white;
  background-color: $extra-light-gray;
  height: 50px;
  width: 100%;
  border-radius: 3px;
  border: none;

  &:focus,
  &:active {
    border-bottom: 2px solid $dark-green;
  }

  &.invalid {
    background-color: $extra-light-red;
    border-bottom: 2px solid $light-red;
  }
}

.validation-message {
  margin-top: 10px;
}
</style>
