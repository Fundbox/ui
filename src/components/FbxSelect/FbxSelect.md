# FbxSelect

## Basic usage

- Render as many `option` tags within `fbx-select` component
- Set validations by passing them via `validations` property
- Get a value by binding your data property to `v-model`

```html
<fbx-select validations="required" v-model="selectedValue" />
```
