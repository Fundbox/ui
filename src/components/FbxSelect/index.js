import FbxSelect from './FbxSelect.vue'
import { registerComponents, vueUse } from '../../utils/plugins'

const components = {
  FbxSelect
}

const VuePlugin = {
  install(Vue) {
    registerComponents(Vue, components)
  }
}

vueUse(VuePlugin)

export default VuePlugin
