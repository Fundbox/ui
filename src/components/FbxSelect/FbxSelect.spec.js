import { shallowMount, createLocalVue } from '@vue/test-utils'
import VeeValidate from 'vee-validate'
import FbxSelect from './FbxSelect.vue'

const option1 = 'Option 1'
const defaultSlot = {
  default: `
    <option>Default</option>
    <option value="${option1}">${option1}</option>
  `
}

const localVue = createLocalVue()
localVue.use(VeeValidate, {
  events: 'change'
})

describe('FbxSelect', () => {
  describe('snapshots', () => {
    it('renders the default correctly', async() => {
      const wrapper = shallowMount(FbxSelect, {
        slots: defaultSlot,
        attrs: {
          name: 'select',
        },
        // running async test because of vee-validate and @vue/test-utils
        // Here are the reffs:
        // https://github.com/baianat/vee-validate/issues/1267
        // https://github.com/baianat/vee-validate/issues/1432
        sync: false,
        localVue
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders validations correctly', async() => {
      const validationMessage = 'This field is required'
      const wrapper = shallowMount(FbxSelect, {
        slots: defaultSlot,
        attrs: {
          name: 'select'
        },
        props: {
          validations: 'required'
        },
        mocks: {
          errors: {
            has: () => true,
            first: () => validationMessage,
          }
        },
        sync: false,
        localVue
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.html()).toContain(validationMessage)
    })
  })

  describe('initialization', () => {
    it('sets correct value by passing value attr', async() => {
      const wrapper = shallowMount(FbxSelect, {
        slots: defaultSlot,
        attrs: {
          name: 'select',
          value: option1,
        },
        sync: false,
        localVue
      })
      await wrapper.vm.$nextTick()
      // promise.then(() => {
      const select = wrapper.find('select')
      const options = select.findAll('option')
      const selectedOption = options.at(select.element.selectedIndex)
      expect(selectedOption.text()).toEqual(option1)
      //   done()
      // })
    })
  })

  describe('events', () => {
    it('calls input event with selected value on value select', () => {
      const mockInput = jest.fn()
      const wrapper = shallowMount(FbxSelect, {
        slots: defaultSlot,
        attrs: {
          name: 'select'
        },
        listeners: {
          input: mockInput,
        },
        sync: false,
        localVue
      })
      const options = wrapper.find('select').findAll('option')
      options.at(1).setSelected()
      expect(mockInput).toHaveBeenCalledTimes(1)
      expect(mockInput).toHaveBeenCalledWith(option1)
    })

    it('calls regular events of select', () => {
      const mockClick = jest.fn()
      const wrapper = shallowMount(FbxSelect, {
        slots: defaultSlot,
        attrs: {
          name: 'select'
        },
        listeners: {
          click: mockClick,
        },
        sync: false,
        localVue
      })
      const select = wrapper.find('select')
      select.trigger('click')
      expect(mockClick).toHaveBeenCalledTimes(1)
    })
  })
})
