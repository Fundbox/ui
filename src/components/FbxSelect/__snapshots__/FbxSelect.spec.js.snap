// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FbxSelect snapshots renders the default correctly 1`] = `
<div><select name="select" class="fbx-select" aria-required="false" aria-invalid="false">
    <option>Default</option>
    <option value="Option 1">Option 1</option>
    <option value="Option 2">Option 2</option>
  </select>
  <!---->
</div>
`;

exports[`FbxSelect snapshots renders validations correctly 1`] = `
<div><select name="select" class="fbx-select invalid" aria-required="false" aria-invalid="false">
    <option>Default</option>
    <option value="Option 1">Option 1</option>
    <option value="Option 2">Option 2</option>
  </select>
  <fbx-validation-message-stub class="validation-message">This field is required</fbx-validation-message-stub>
</div>
`;
